### DISCUSSION TITLE
SW3 Will Not Run Script

### DISCUSSION THREAD

**Discussion by lindelux** [PiSupply/PiJuice#1055] [Q&A] [answered]
I am using PiJuice in cli mode on a Pi 3 with Bookworm.

1. I have a python script to print ("hello"). 
2. It is set to executable. 
3. It is assigned to FUNC1 as /home/pi/Scripts/test.py. 
4. I have set function for PRESS on SW3 to USER_FUNC1. 
5. Running test,py outside of pijuice works OK.

When SW3 is pressed nothing happens. What have I overlooked?

**Comment by tvoverbeek** [comment]
If you look at the figure of the top of the board in the hardware section you see the order of the switches is SW1, SW3, SW2.
So SW3 is the one in the middle

**Reply by lindelux** [reply]
Thanks for responding. Yes, 1, 3, 2. I am selecting the correct switch, the middle one.

**Comment by tvoverbeek** [comment]
Who his the owner of /home/pi/Scripts/test.py?
If the script owner is root, the script will not be executed.
Also the owner of the script must belong to the pijuice group.
what is the output fo `ls -l /home/pi/Scripts/test.py`?

**Reply by lindelux** [reply]
-rwxrwxrwx 1 pijuice pijuice  203 Jan  7 09:59 test.py

**Comment by tvoverbeek** [comment]
The script is executed as part of the pijuice service (/usr/bin/pijuice_sys.py).
The output from print statements does **not** go to the console.
Since the print output is buffered, add a `flush=True` argument to the print statement.
To see the output use `systemctl stat pijuice`.
An other thing to check is the directory protections along the path (/home/pi and /home/piScripts in your case).
All should have group protection `r-x`. If not, the pijuice service cannot execute the script.
My own home directory (/home/tvo) did have group protection `---`
Here is the output from `systemctl status pijuice` after tne group protection correction:
```
tvo@rpi5-64:~/Scripts $ systemctl status pijuice
● pijuice.service - PiJuice status service
     Loaded: loaded (/lib/systemd/system/pijuice.service; disabled; preset: enabled)
     Active: active (running) since Mon 2024-01-08 09:57:36 CET; 25min ago
   Main PID: 2926 (pijuice_sys.py)
      Tasks: 1 (limit: 4446)
        CPU: 537ms
     CGroup: /system.slice/pijuice.service
             └─2926 /usr/bin/python3 /usr/bin/pijuice_sys.py

Jan 08 09:58:12 rpi5-64 sudo[2981]:  pijuice : PWD=/var/lib/pijuice ; USER=pijuice ; COMMAND=/home/tvo/Scripts/test.py PRESS SW3
Jan 08 09:58:12 rpi5-64 sudo[2981]: pam_unix(sudo:session): session opened for user pijuice(uid=1001) by (uid=1001)
Jan 08 09:58:12 rpi5-64 pijuice_sys.py[2983]: test
Jan 08 09:58:12 rpi5-64 sudo[2981]: pam_unix(sudo:session): session closed for user pijuice
tvo@rpi5-64:~/Scripts $ 
```

**Reply by lindelux** [reply]
First I want to say think you for your help!

Here is my script with the flush=True added to the print statement.:

#!/usr/bin/python3
from time import sleep
from pijuice import PiJuice # Import pijuice module
pijuice = PiJuice(1, 0x14) # Instantiate PiJuice interface object
print ("hello", flush=True)

Directory protections:
-rwxrwxrwx  1 pi pi  192 Jan  8 17:27 test.py

After selecting SW3 the output of  output use systemctl stat pijuice give the following;

● pijuice.service - PiJuice status service
     Loaded: loaded (/lib/systemd/system/pijuice.service; enabled; preset: enabled)
     Active: active (running) since Mon 2024-01-08 17:24:59 -03; 7min ago
   Main PID: 511 (pijuice_sys.py)
      Tasks: 1 (limit: 758)
        CPU: 1.465s
     CGroup: /system.slice/pijuice.service
             └─511 /usr/bin/python3 /usr/bin/pijuice_sys.py

It's like pijuice doesn't see when SW3 is selected. I'm stumped.

-rw------- 1 pijuice pijuice 1652 Jan  8 17:32 pijuice_config.JSON

UPDATE
I tried this script https://learn.pi-supply.com/make/how-to-run-a-user-script-with-the-pijuice-power-events/
and it didn't work either.  There must be some setting or lack thereof that is preventing any script from running.

**Comment by tvoverbeek** [comment]
What is the output from `ls -l home/pi`?
If it is `-rwx-------` then your scripts woll not run.
It should be `-rwxr-x----`or `-rwxr-xr-x-`
It seems recent version of the imager creates the home directory with group protection `---`

**Reply by lindelux** [reply]
scripts $ ls -l
total 20
-rw-r--r-- 1 pi pi  821 Jan  8 19:02 LED_green.py
-rwxr-xr-x 1 pi pi  942 Jan  8 18:53 LED.py
-rwxr-xr-x 1 pi pi  921 Jan  8 18:29 shutdown.py
-rwxr-xr-x 1 pi pi 1270 Jan  8 19:42 status.py
-rwxr-xr-x 1 pi pi  192 Jan  8 17:27 test.py

**Comment by tvoverbeek** [comment]
Sorry, I meant `ls -ld /home/pi`

**Reply by lindelux** [reply]
pi@batt:~ $ ls -ld /home/pi
drwxr-x--- 9 pi pi 4096 Jan 10 19:05 /home/pi

**Reply by lindelux** [reply]
I wrote an other script to turn on LED D2. When I run "python LED_on.py" from the terminal then D2 turns green. I set this script to USER_FUNC5 and assigned it to SW3. Selecting SW3 does not run the script and turn on the LED. At this point is it possible the pijuice is faulty?

**Reply by lindelux** [reply]
after pressing SW3 and running sudo systemctl stat this is what I get; no sign of the script running !!

pi@batt:~ $ sudo systemctl status pijuice
● pijuice.service - PiJuice status service
     Loaded: loaded (/lib/systemd/system/pijuice.service; enabled; preset: enabled)
     Active: active (running) since Wed 2024-01-10 19:05:42 -03; 17min ago
   Main PID: 501 (pijuice_sys.py)
      Tasks: 1 (limit: 758)
        CPU: 2.891s
     CGroup: /system.slice/pijuice.service
             └─501 /usr/bin/python3 /usr/bin/pijuice_sys.py

Jan 10 19:05:42 batt systemd[1]: Started pijuice.service - PiJuice status service.
Jan 10 19:05:51 batt pijuice_sys.py[501]: RTC os-support OK

**Comment by tvoverbeek** [comment]
The protection of /home/pi is OK (group has read and execute access).

Can you try the following: disable the pijuice service and run pijuice_sys.py manually:
```
sudo systemctl disable pijuice
sudo stop pijuice
sudo -u pijuice /usr/bin/python3 /usr/bin/pijuice_sys.py
```
Press SW3 and see if you get any error messages
Stop pijuice_sys with CTRL-C.

Then restart the service:
```
sudo systemctl enable pijuice
sudo systemctl start pijuice
```

**Reply by lindelux** [reply]
pi@batt:~ $ sudo systemctl disable pijuice
Removed "/etc/systemd/system/multi-user.target.wants/pijuice.service".
pi@batt:~ $ sudo stop pijuice
sudo: stop: command not found
pi@batt:~ $ sudo -u pijuice /usr/bin/python3 /usr/bin/pijuice_sys.py
RTC os-support OK
^CTraceback (most recent call last):
  File "/usr/bin/pijuice_sys.py", line 425, in <module>
    main()
  File "/usr/bin/pijuice_sys.py", line 421, in main
    time.sleep(1)
KeyboardInterrupt

pi@batt:~ $ sudo systemctl enable pijuice
Created symlink /etc/systemd/system/multi-user.target.wants/pijuice.service → /lib/systemd/system/pijuice.service.
pi@batt:~ $ sudo systemctl start pijuice

**Comment by tvoverbeek** [comment]
There is an error in my instructions above.
`sudo stop pijuice` does not work as you noticed.
It should be `sudo systemctl stop pijuice`.
WIth the erroneous command you had 2 copies of pijuice_sys.py running.
Could you try again (after a reboot, just to be sure) and see what messages you get after pressing SW3.
Try also SW1 and SW3.

**Reply by lindelux** [reply]
I followed your instruction as follows:

sudo systemctl disable pijuice
sudo sudo systemctl stop pijuice.
sudo -u pijuice /usr/bin/python3 /usr/bin/pijuice_sys.py

the above displays "RTC os-support OK" on the screen.

I pressed  SW1, SW2 and SW3 and received no errors, there was no output to the screen whatsoever.

Stop pijuice_sys with CTRL-C.

Then restart the service:

sudo systemctl enable pijuice
sudo systemctl start pijuice

**Comment by tvoverbeek** [comment]
Strange ..
What have you defined as button SW3 function? USER_FUNCTION5?
And for the USER_FUNCTION5 ? LED_ON.py?
Screenshots preferred.
Note thst in the USER-FUNCTION5 definition you have to specify the full path to the function.
Also LED_ON.py has to be executable and the first line of the function `!/usr/bin/python3`.
To test it try `sudo -u pijuice /home/pi/LED_ON.py` assuming /ome/pi/LED_ON.py is the full path to the script.

**Reply by lindelux** [reply]
All the scripts are executable.
![Screenshot from 2024-01-13 12-07-29](https://github.com/PiSupply/PiJuice/assets/64298178/0bb1c8fe-f8d8-4e73-a8d0-c6ea54660914)

User scripts:
![User_scripts](https://github.com/PiSupply/PiJuice/assets/64298178/093c730c-50cd-42af-b496-3dcb20daab37)

Settings for SW3:
![Screenshot from 2024-01-13 12-10-37](https://github.com/PiSupply/PiJuice/assets/64298178/0b416d25-25a7-4cff-b7c3-b302424b2701)

I tried your suggestion "sudo -u pijuice /home/pi/LED_on.py" and it did not work. I found I had a space above the shebang in the script so I removed the space and then "sudo -u pijuice /home/pi/LED_on.py" and "sudo -u pijuice /home/pi/LED_off.py" worked fine.

Still when I press SW3 nothing happens so I tried your previous instructions and got no messages.

![Screenshot from 2024-01-13 12-16-34](https://github.com/PiSupply/PiJuice/assets/64298178/f9e02c79-54ea-49a2-a2f5-d999941bd512)

**Comment by tvoverbeek** [comment, marked as answer]
Thanks for the screenshots.
Your SW3 settings do not work.
For SINGLE_PRESS you specify USER_FUNC5 followed by a 0 parameter,
The parameter specifies a timeout in milliseconds wihin the button should be released to trigger the SINGLE_PRESS event. Try a parameter value of 500. Specifying 0 (as in your screenshot) effectively disables the  SINGLE_PRESS event.
See also the Software Readme sections on button settings.

**Reply by lindelux** [reply]
OK now it's working, I set the parameter to 1000 and it have it set so SW3 turns on the LED and SW2 turns off the LED.
I really appreciate your assistance. I hope the information in this posting will help others.

MANY THANKS !!
